eclass/secureboot.eclass

```diff
@@ -0,0 +1,179 @@
+
+if [[ -z ${_SECUREBOOT_ECLASS} ]]; then
+_SECUREBOOT_ECLASS=1
+
+# @FUNCTION: _secureboot_die_if_unset
+# @INTERNAL
+# @DESCRIPTION:
+# If USE=secureboot is enabled die if the required user variables are unset
+# and die if the keys can't be found.
+_secureboot_die_if_unset() {
+	use secureboot || return
+	debug-print-function ${FUNCNAME[0]} "${@}"
+	if [[ -z ${SECUREBOOT_SIGN_KEY} || -z ${SECUREBOOT_SIGN_CERT} ]]; then
+		die "USE=secureboot enabled but SECUREBOOT_SIGN_KEY and/or SECUREBOOT_SIGN_CERT not set."
+	fi
+	if [[ ! ${SECUREBOOT_SIGN_KEY} == pkcs11:* && ! -f ${SECUREBOOT_SIGN_KEY} ]]; then
+		die "SECUREBOOT_SIGN_KEY=${SECUREBOOT_SIGN_KEY} not found"
+	fi
+	if [[ ! -f ${SECUREBOOT_SIGN_CERT} ]];then
+		die "SECUREBOOT_SIGN_CERT=${SECUREBOOT_SIGN_CERT} not found"
+	fi
+}
+
+# @FUNCTION: secureboot_pkg_setup
+# @DESCRIPTION:
+# Checks if required user variables are set before starting the build
+secureboot_pkg_setup() {
+	use secureboot || return
+	debug-print-function ${FUNCNAME[0]} "${@}"
+	# If we are merging a binary then the files in this binary
+	# are already signed, no need to check the variables.
+	if [[ ${MERGE_TYPE} != binary ]]; then
+		_secureboot_die_if_unset
+	fi
+}
+
+# @FUNCTION: secureboot_sign_efi_file
+# @USAGE: <input file> <output file>
+# @DESCRIPTION:
+# Sign a file using sbsign and the requested key/certificate.
+# If the file is already signed with our key then skip.
+secureboot_sign_efi_file() {
+	use secureboot || return
+	debug-print-function ${FUNCNAME[0]} "${@}"
+	local input_file=${1}
+	local output_file=${2}
+
+	_secureboot_die_if_unset
+
+	if has_version app-crypt/sbsigntools; then
+		ebegin "Signing ${input_file}"
+		local return=1
+		if sbverify "${input_file}" --cert "${SECUREBOOT_SIGN_CERT}" &> /dev/null; then
+			ewarn "${input_file} already signed, skipping"
+			return=0
+		else
+			local args=(
+				"--key=${SECUREBOOT_SIGN_KEY}"
+				"--cert=${SECUREBOOT_SIGN_CERT}"
+			)
+			if [[ ${SECUREBOOT_SIGN_KEY} == pkcs11:* ]]; then
+				args+=( --engine=pkcs11 )
+			fi
+
+			sbsign "${args[@]}" "${input_file}" --output "${output_file}"
+			return=${?}
+		fi
+		eend ${return} || die -n "Signing ${input_file} failed"
+	else
+		die "secureboot signing requested but app-crypt/sbsigntools not installed"
+	fi
+}
+
+# @FUNCTION: secureboot_auto_sign
+# @USAGE: [--in-place | --ext <file extension>]
+# @DESCRIPTION:
+# Automatically discover and sign efi files in the image directory.
+#
+# By default signed files gain the .signed extension. If the --in-place
+# argument is given the efi files are replaced with a signed version in place.
+# The --ext argument allows for overriding the default .signed file extension.
+secureboot_auto_sign() {
+	use secureboot || return
+	debug-print-function ${FUNCNAME[0]} "${@}"
+
+	[[ ${EBUILD_PHASE} == install ]] ||
+		die "${FUNCNAME[0]} can only be called in the src_install phase"
+
+	local path=${ED}
+	local -a efi_execs
+	[[ -d ${path} ]] && mapfile -td '' efi_execs < <(
```

Comment:
Is there a valid case for calling this function when `${ED}` does not exist?

Reply:
fwiw it allows to give the error just below rather than have `find` itself ungracefully fail on missing path

Reply:
Albeit I guess this is partially copied from linux-mod-r1, except in linux-mod-r1 you could specify the path rather than it always be `${ED}`

Reply:
Yeah this was copied from `linux-mod-r1.eclass. But this check isn't really necessary in our case indeed.